src/core/api.proto

```diff
@@ -94,6 +94,55 @@ message InferRequestHeader
     //@@       for tensors with a non-fixed-size datatype (like STRING).
     //@@
     uint64 batch_byte_size = 3;
+
+    //@@      .. cpp:var:: message SharedMemoryElement
+    //@@
+    //@@         The Meta-data for the Input shared memory to read element from.
+    //@@
+    message SharedMemoryElement
+    {
+      //@@    .. cpp:var:: key_t shm_key
+      //@@
+      //@@     The unique key used to refer to a specific location in shared
+      //@@     memory. Optional for tensors. Required to read tensors from
+      //@@     shared memory. If provided, tensor data read from this location.
+      //@@
+      key_t shm_key = 1;
+
+      //@@    .. cpp:var:: uint64 offset
+      //@@
+      //@@     The offset for the each element.
+      //@@     element0 start = offset0, end = offset0 + size;
+      //@@
+      uint64 offset = 2;
+
+      //@@    .. cpp:var:: uint64 byte_size
+      //@@
+      //@@     Size of the element in the batch of the input tensor, in bytes.
+      //@@
+      uint64 byte_size = 3;
+    }
+
+    //@@    .. cpp:var:: message SharedMemoryIO
+    //@@
+    //@@       Meta-data for an input batch to be read from shared memory.
+    //@@
+    message SharedMemoryIO
+    {
+      //@@      .. cpp:var:: SharedMemoryElement element (repeated)
+      //@@
+      //@@         No. of elements required is number of elements in the batch.
+      //@@
+      repeated SharedMemoryElement element = 1;
+    }
+
+    //@@    .. cpp:var:: SharedMemoryIO shm_io
+    //@@
+    //@@       Unique identity of the location in shared memory to read input
```

Comment:
This should say optional. And then just something like "location of the input tensor in shared memory". Don't talk about responsibilities here.